specs/agents/tracing-instrumentation-aws-lambda.md

```diff
@@ -0,0 +1,158 @@
+
+
+### S3
+Lambda functions that are triggered by S3 accept an `event` input that may contain one ore more `S3 event notification records` in the `event.records` array. We cannot (automatically) wrap the processing of the individual records that are sent as a batch of S3 event notification records with a single `event`. 
+
+Thus, in case that an S3 `event` contains **exactly one** `S3 event notification record`, the agents must apply the following, S3-specific retrieval of information. Otherwise, the agents should apply the [Generic Lambda Instrumentation](generic-lambda-instrumentation) as desribed above.
+
+In addition the following fields should be set for Lambda functions triggered by S3:
+Field | Value | Description 
+---   | ---   | ---
+`faas.trigger.type` | `datasource` | Constant value for message based triggers.
+`faas.trigger.name` | e.g. `mybucket` | Use `record.s3.bucket.name`
+`faas.trigger.id` | e.g. `arn:aws:s3:::mybucket/ObjectCreated:Put` | Use `record.s3.bucket.arn`
+`faas.trigger.region` | e.g. `us-east-1` | Use `record.awsRegion`
+`faas.trigger.version` | e.g. `2.1` | Use `record.eventVersion`
+`faas.execution` | e.g. `arn:aws:s3:::mybucket/ObjectCreated:Put` | Format: `${record.s3.bucket.arn}/${record.eventName}`
+
+
+## Init spans
+With both API Gateway versions the `event` objects (may) contain a timestamp of the original API Gateway request. The difference between the API Gateway timestamp and the actual start timestamp of the Lambda function is a good estimate for the initialization phase of the lambda function. 
+The agents should use this information (if available) to adapt the transaction and spans structure of the Lambda invocation to represent the initialization time:
+
+```
+...
+  [XXXXXXXXXXXXX POST api.gateway XXXXXXXXXXXXXXXXXXXXX]   // client calling the API Gateway
+     [XXXXXXXXXX MyLambdaFunction XXXXXXXXXXXXXXXXXXXX]    // Lambda Transaction
+     [XXX Lambda init XXXXX]                               // Init span
+                            [XXXX Lambda handler XXXX]      // span for the handler method
+                                    [XXXXXXXXXX]           // any Lambda subspans
+     ^                      ^
+     |                      |
+API Gateway               Timestamp of the
+RequestTimestamp          handler method start
+```
+
+To achieve the above transaction / span structure we need to backdate the begin of the created transaction using the API Gateway timestamp. We create a artificial span `Lambda init` using the API Gateway timestamp as start and the handler method timestamp as end. The `Lambda handler` span represents the actual execution of the Lambda handler method. 
+
+### Retrieving the API Gateway V1 timestamp
```

Comment:
I'm not sure this is valuable. From everything I've read, the v2 API Gateway is faster, cheaper, and more feature-rich. I don't think we need to put in the work to support v1 in this way, especially as seconds-level granularity is largely useless in an APM context. I don't have stats to back up this gut feeling, but I think we could safely ignore the v1 issues.

Reply:
agree, will remove this